src/main/org/apache/tools/ant/taskdefs/modules/Jmod.java

```diff
@@ -0,0 +1,1282 @@
+            && modulePath == null) {
+
+            throw new BuildException(
+                "hashModulesPattern requires a module path, since "
+                + "it will generate hashes of the other modules which depend "
+                + "on the module being created.",
+                getLocation());
+        }
+
+        checkDirPaths();
+
+        Path[] dependentPaths = {
+            classpath,
+            modulePath,
+            commandPath,
+            configPath,
+            headerPath,
+            legalPath,
+            nativeLibPath,
+            manPath,
+        };
+        Union allResources = new Union(getProject());
+        for (Path path : dependentPaths) {
+            if (path != null) {
+                for (String entry : path.list()) {
+                    File entryFile = new File(entry);
+                    if (entryFile.isDirectory()) {
+                        log("Will compare timestamp of all files in "
+                            + "\"" + entryFile + "\" with timestamp of "
+                            + jmodFile, Project.MSG_VERBOSE);
+                        FileSet fileSet = new FileSet();
+                        fileSet.setDir(entryFile);
+                        allResources.add(fileSet);
+                    } else {
+                        log("Will compare timestamp of \"" + entryFile + "\" "
+                            + "with timestamp of " + jmodFile,
+                            Project.MSG_VERBOSE);
+                        allResources.add(new FileResource(entryFile));
+                    }
+                }
+            }
+        }
+
+        ResourceCollection outOfDate =
+            ResourceUtils.selectOutOfDateSources(this, allResources,
+                new MergingMapper(jmodFile.toString()),
+                getProject(),
+                FileUtils.getFileUtils().getFileTimestampGranularity());
+
+        if (outOfDate.isEmpty()) {
+            log("Skipping jmod creation, since \"" + jmodFile + "\" "
+                + "is already newer than all files in paths.",
+                Project.MSG_VERBOSE);
+            return;
+        }
+
+        Collection<String> args = buildJmodArgs();
+
+        try {
+            log("Deleting " + jmodFile + " if it exists.", Project.MSG_VERBOSE);
+            Files.deleteIfExists(jmodFile.toPath());
+        } catch (IOException e) {
+            throw new BuildException(
+                "Could not remove old file \"" + jmodFile + "\": " + e, e,
+                getLocation());
+        }
+
+        ToolProvider jmod = ToolProvider.findFirst("jmod").orElseThrow(
+            () -> new BuildException("jmod tool not found in JDK.",
+                getLocation()));
+
+        log("Executing: jmod " + String.join(" ", args), Project.MSG_VERBOSE);
+
+        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+        ByteArrayOutputStream stderr = new ByteArrayOutputStream();
```

Comment:
Hi Craig, like Stefan already said, this patch is a really good addition to Ant. Thank you for that.

The only question/suggestion I have is, should we instead just pass `System.out` and `System.err` instead of creating a `ByteArrayOutputStream`? That way the actual output and errors are logged instead of we deciding to print it only when the tool execution fails.

Reply:
That was what I tried first, but the output and error were never showing up at all in unit tests.  (Also, my observation has been that the tool produces no output when it runs successfully.)

Reply:
Test setup captures streams so it may become difficult. Would using `LogOutputStreams` do which would send the output directly to An't logging system?